File: trabalho-sistema-eleitoral-brasileiro-java/TP1/src/Read.java
package TP1;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.LinkedList;
import java.util.List;
import java.util.Scanner;

public class Read {
	/**
  	 * @brief Esse método implementa a leitura do arquivo de partidos políticos.
  	 * As informações sobre os partidos são armazenadas em uma lista encadeada.
  	 * @param:
 	 * 		@fileName Nome do arquivo que será lido.
  	 */
     public static List<PoliticParty> readPoliticParty(String fileName) throws FileNotFoundException {
        List<PoliticParty> partys = new LinkedList<PoliticParty>();
        try {
        	Scanner input = new Scanner(new File(fileName), "utf-8");
			input.nextLine(); 
			
			while (input.hasNextLine()) {
			    String line = input.nextLine();
                String linePartys[] = line.split(",");
                
                int number = Integer.parseInt(linePartys[0]);
                int legendVotes = Integer.parseInt(linePartys[1]);
                String name = linePartys[2];
                String abreviation = linePartys[3];
                
				PoliticParty p = new PoliticParty(name, abreviation, number, legendVotes);
				partys.add(p);
			}
			
			input.close();
		
        } catch (FileNotFoundException e) {
            System.out.println("An Error has ocurred! FileNotFoundException");
        }
        
        return partys;
    }
     
 	/**
   	 * @brief Esse método implementa a leitura do arquivo de candidatos.
   	 * As informações sobre os candidatos são armazenadas em uma lista encadeada.
   	 * @param:
  	 * 		@fileName Nome do arquivo que será lido.
   	*/
    public static List<Candidate> readCandidate(String fileName) throws FileNotFoundException {
        List<Candidate> candidates = new LinkedList<Candidate>();
        
        try {
        	Scanner input = new Scanner(new File(fileName), "utf-8");
			input.nextLine(); 
			
			while (input.hasNextLine()) {
			    String line = input.nextLine();
                String lineParts[] = line.split(",");
                
                int balBoxNumber = Integer.parseInt(lineParts[0]);
                int nominalVotes = Integer.parseInt(lineParts[1]);
                String situation = lineParts[2];
                String name = lineParts[3];
                String balBoxName = lineParts[4];
                char gender = lineParts[5].charAt(0);
                String birth = lineParts[6];
                String voteDestination = lineParts[7];
                int partyNumber = Integer.parseInt(lineParts[8]);

				Candidate p = new Candidate(name, gender, birth, balBoxName, balBoxNumber, nominalVotes, situation, partyNumber, voteDestination);
				candidates.add(p);
			}
			
			input.close();
		
        } catch (FileNotFoundException e) {
        	System.out.println("An Error has ocurred! FileNotFoundException");
        }
        
        return candidates;
    }
}

File: trabalho-sistema-eleitoral-brasileiro-cpp/TP2/src/Read.cpp
#include "../include/Read.h"
// #include "Read.h"

vector<PoliticParty*> read_politic_party(const string& fileName) {
    vector<PoliticParty*> partys;

    try {
        ifstream fin(fileName);  
        
        if(!fin.good()) {
            fin.close();
            file_not_found_exception();
        }

        string line;
        getline(fin, line);
        while(!fin.eof()) {
            getline(fin, line);
    
            Tokenizer tokenizer(line, ',');
            vector<string> split = tokenizer.remaining();
            
            if(split.size() < 4) continue;

            int number = stoi(split.at(0));
            int legendVotes = stoi(split.at(1));
            string name = split.at(2);
            string abreviation = split.at(3);

            PoliticParty* p = new PoliticParty(name, abreviation, number, legendVotes);
            partys.push_back(p);
        }

        fin.close();

    } catch(Exception e) {
        cerr << e << endl;
        exit(1);
    }

    // catch(exception& e) {
    //     cerr << "Exception: [" << e.what() << "]";
    // }

    // catch(ios_base::failure& e) {
    //     cout << "Exception: [";
    //     cerr << e.what();
    //     cout << "]" << endl;
    // }

    return partys;
}

vector<Candidate*> read_candidate(const string& fileName) {
    vector<Candidate*> candidates;

    try {
        ifstream fin(fileName);

        if(!fin.good()) {
            fin.close();
            file_not_found_exception();
        }

        string line;
        getline(fin, line);

        while(!fin.eof()) {
            getline(fin, line);
    
            Tokenizer tokenizer(line, ',');
            vector<string> split = tokenizer.remaining();
            
            if(split.size() < 9) continue;

            int balboxNumber = stoi(split.at(0));
            int nominalVotes = stoi(split.at(1));
            string situation = split.at(2);
            string name = trim(split.at(3));
            string balboxName = trim(split.at(4));
            char gender = split.at(5).at(0);
            string birth = split.at(6);
            string voteDestination = split.at(7);
            int partyNumber = stoi(split.at(8));

            Candidate* p = new Candidate(name, gender, birth, balboxName, balboxNumber, nominalVotes, situation, partyNumber, voteDestination);
            candidates.push_back(p);
        }

        fin.close();

    } catch(Exception e) {
        cerr << e << endl;
        exit(1);
    }

    return candidates;
}
